test/companies/get-company-by-registration-number.api-test.ts

```diff
@@ -0,0 +1,144 @@
+import { RandomValueGenerator } from '@ukef-test/support/generator/random-value-generator';
+import nock from 'nock';
+
+describe('GET /companies?registrationNumber=', () => {
+  let api: Api;
+
+  const valueGenerator = new RandomValueGenerator();
+
+  const {
+    companiesHousePath,
+    mdmPath,
+    getCompanyCompaniesHouseResponse,
+    getCompanyCompaniesHouseMalformedAuthorizationHeaderResponse,
+    getCompanyCompaniesHouseInvalidAuthorizationResponse,
+    getCompanyCompaniesHouseNotFoundResponse,
+    getCompanyResponse,
+  } = new GetCompanyGenerator(valueGenerator).generate({
+    numberToGenerate: 1,
+    registrationNumber: '00000001',
+  });
+
+  const getMdmPath = (registrationNumber: string) => `/api/v1/companies?registrationNumber=${encodeURIComponent(registrationNumber)}`;
+
+  beforeAll(async () => {
+    api = await Api.create();
+  });
+
+  afterAll(async () => {
+    await api.destroy();
+  });
+
+  afterEach(() => {
+    nock.abortPendingRequests();
+    nock.cleanAll();
+  });
+
+  // MDM auth tests
+  withClientAuthenticationTests({
+    givenTheRequestWouldOtherwiseSucceed: () => {
+      requestToGetCompanyByRegistrationNumber(companiesHousePath).reply(200, getCompanyCompaniesHouseResponse);
+    },
+    makeRequestWithoutAuth: (incorrectAuth?: IncorrectAuthArg) => api.getWithoutAuth(mdmPath, incorrectAuth?.headerName, incorrectAuth?.headerValue),
+  });
+
+  it('returns a 200 response with the company if it is returned by the Companies House API', async () => {
+    requestToGetCompanyByRegistrationNumber(companiesHousePath).reply(200, getCompanyCompaniesHouseResponse);
+
+    const { status, body } = await api.get(mdmPath);
+
+    expect(status).toBe(200);
+    expect(body).toStrictEqual(getCompanyResponse);
+  });
+
+  it.each([
+    {
+      registrationNumber: valueGenerator.stringOfNumericCharacters({ length: 6 }),
+      validationError: 'registrationNumber must be longer than or equal to 7 characters',
+    },
+    {
+      registrationNumber: valueGenerator.stringOfNumericCharacters({ length: 9 }),
+      validationError: 'registrationNumber must be shorter than or equal to 8 characters',
+    },
+    {
+      registrationNumber: '0A000001',
+      validationError: 'registrationNumber must match /^(([A-Z]{2}|[A-Z]\\d|\\d{2})(\\d{5,6}|\\d{4,5}[A-Z]))$/ regular expression',
+    },
+  ])(`returns a 400 response with validation errors if postcode is '$registrationNumber'`, async ({ registrationNumber, validationError }) => {
+    const { status, body } = await api.get(getMdmPath(registrationNumber));
+
+    expect(status).toBe(400);
+    expect(body).toMatchObject({
+      error: 'Bad Request',
+      message: expect.arrayContaining([validationError]),
+      statusCode: 400,
+    });
+  });
+
+  it(`returns a 500 response if the Companies House API returns a 400 response containing the error string 'Invalid Authorization header'`, async () => {
```

Comment:
Can we have api-test for 404 response.

Reply:
Can we also add `429` please.

Reply:
Added `404` test. For a `429` from Companies House API, I think we should return a `500` from our endpoint, as a `429` could be misleading in that it suggests we have hit the MDM rate-limit. I think our rate-limit should be sufficiently low to prevent `429`s from 3rd party services, and that it should be considered an internal error (i.e., we should decrease our rate-limit) rather than a client error if we receive a `429` from CH API. As we don't specifically handle a `429` response from CH API and we can't add tests to cover every possible response from CH API, I feel that this case is covered by the final `'returns a 500 response if the request to the Companies House API returns an unhandled error response'` test case. However, I'm happy to add a `429` from CH API = `500` from our endpoint test case if you feel it is needed.